aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -0,0 +1,209 @@
+
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ gesv ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+template<typename scalar_t>
+static void apply_gesv(Tensor& b, Tensor& A, std::vector<int64_t>& infos) {
+#ifndef USE_LAPACK
+  AT_ERROR("gesv: LAPACK library not found in compilation");
+#endif
+  auto A_data = A.data<scalar_t>();
+  auto b_data = b.data<scalar_t>();
+  auto A_mat_stride = matrixStride(A);
+  auto b_mat_stride = matrixStride(b);
+
+  auto batch_size = batchCount(A);
+  auto n = A.size(-2);
+  auto nrhs = b.size(-1);
+
+  auto ipiv = at::empty({n}, b.type().toScalarType(kInt));
+
+  for (int64_t i = 0; i < batch_size; i++) {
+    int info;
+    scalar_t* A_working_ptr = &A_data[i * A_mat_stride];
+    scalar_t* b_working_ptr = &b_data[i * b_mat_stride];
+    lapackGesv<scalar_t>(n, nrhs, A_working_ptr, n, ipiv.data<int>(), b_working_ptr, n, &info);
+    infos[i] = info;
+    if (info != 0) {
+      return;
+    }
+  }
+}
+
+// These utilities are specified in LinearAlgebraUtils.h
+LINALG_HELPER_2_ARGS(gesv, self, A, cpu)
+
+// Supports arbitrary batch dimensions for self and A
+std::tuple<Tensor,Tensor> gesv(const Tensor& self, const Tensor& A) {
+  if (self.dim() <= 2 && A.dim() <= 2) {
+    // TODO: #7102: It's not necessary to have gesv (single) bindings for both
+    // TH and ATen. We should remove the TH gesv bindings, especially
+    // since the lapackGesv function is already in ATen.
+    return at::_gesv_single(self, A);
+  }
+
+  gesvCheckInputs(self, A);
+
+  // broadcast the batch dimensions of self and A.
+  IntList self_batch_sizes(self.sizes().data(), self.ndimension() - 2);
+  IntList A_batch_sizes(A.sizes().data(), A.ndimension() - 2);
+  std::vector<int64_t> expand_batch_portion = infer_size(self_batch_sizes, A_batch_sizes);
+
+  std::vector<int64_t> self_expand_size({expand_batch_portion});
+  self_expand_size.insert(self_expand_size.end(), { self.size(-2), self.size(-1) });
+
+  std::vector<int64_t> A_expand_size({expand_batch_portion});
+  A_expand_size.insert(A_expand_size.end(), { A.size(-2), A.size(-1) });
+
+  Tensor self_broadcasted  = self.expand(self_expand_size);
+  Tensor A_broadcasted = A.expand(A_expand_size);
+  return at::_gesv_helper(self_broadcasted, A_broadcasted);
+}
+
+std::tuple<Tensor&,Tensor&> gesv_out(Tensor& solution, Tensor& lu, const Tensor& self, const Tensor& A) {
+  AT_CHECK(self.dim() == 2 && A.dim() == 2, 
+           "torch.gesv() with the `out` keyword does not support batching. "
+           "b.dim() (", self.dim(), ") and A.dim() (", A.dim(), ") must both be 2.");
+  return at::_gesv_single_out(solution, lu, self, A);
+}
+
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ inverse ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+template <typename scalar_t>
+static void apply_inverse(Tensor& self, std::vector<int64_t>& infos) {
+#ifndef USE_LAPACK
+  AT_ERROR("inverse: LAPACK library not found in compilation");
+#endif
+  auto self_data = self.data<scalar_t>();
+  auto self_matrix_stride = matrixStride(self);
```

Comment:
Maybe add an assert that self has memory in column major form? (that is the case, right?)